src/D2L.CodeStyle.Analyzers/ApiUsage/Serialization/ReflectionSerializerAnalyzer.cs

```diff
@@ -0,0 +1,314 @@
+			}
+
+			ReflectionSerializerModel model = new ReflectionSerializerModel(
+					ignoreAttributeType: ignoreAttributeType,
+					reflectionSerializerAttributeType: reflectionSerializerAttributeType
+				);
+
+			context.RegisterSymbolAction(
+					c => AnalyzeAttributeSyntax( c, model, (INamedTypeSymbol)c.Symbol ),
+					SymbolKind.NamedType
+				);
+		}
+
+		private static void AnalyzeAttributeSyntax(
+				SymbolAnalysisContext context,
+				ReflectionSerializerModel model,
+				INamedTypeSymbol type
+			) {
+
+			if( type.TypeKind != TypeKind.Class ) {
+				return;
+			}
+
+			if( type.IsAbstract ) {
+				return;
+			}
+
+			if( !model.HasReflectionSerializerAttribute( type, out AttributeData reflectionSerializerAttribute ) ) {
+				return;
+			}
+
+			if( type.IsStatic ) {
+				ReportStaticClass( context, reflectionSerializerAttribute );
+				return;
+			}
+
+			AnalyzeConstructors( context, model, reflectionSerializerAttribute, type );
+			AnalyzeProperties( context, type );
+		}
+
+		private static void AnalyzeConstructors(
+				SymbolAnalysisContext context,
+				ReflectionSerializerModel model,
+				AttributeData reflectionSerializerAttribute,
+				INamedTypeSymbol type
+			) {
+
+			ImmutableArray<IMethodSymbol> constructors = type.InstanceConstructors;
+
+			ImmutableArray<IMethodSymbol> publicConstructors = constructors
+				.Where( c => c.DeclaredAccessibility == Accessibility.Public )
+				.ToImmutableArray();
+
+			if( publicConstructors.IsEmpty ) {
+
+				ReportNoPublicConstructor( context, reflectionSerializerAttribute );
+				return;
+			}
+
+			/*
+			 * The ReflectionSerializer picks the first constructor so just assuming symbols
+			 * align with reflection.  In any event, we will follow up with reporting that 
+			 * multiple constructors exist.
+			 */
+			IMethodSymbol deserializationConstructor = publicConstructors[0];
+			AnalyzeConstructorParameters( context, model, type, deserializationConstructor );
+
+			if( publicConstructors.Length > 1 ) {
+				ReportMultiplePublicConstructors( context, publicConstructors );
+			}
+		}
+
+		private static void AnalyzeConstructorParameters(
+				SymbolAnalysisContext context,
+				ReflectionSerializerModel model,
+				INamedTypeSymbol type,
+				IMethodSymbol constructor
+			) {
+
+			ImmutableArray<IParameterSymbol> constructorParameters = constructor.Parameters;
+			if( constructorParameters.IsEmpty ) {
+				return;
+			}
+
+			ImmutableHashSet<string> serializedPropertyNames = model.GetPublicReadablePropertyNames( type );
+
+			foreach( IParameterSymbol parameter in constructorParameters ) {
+
+				RefKind refKind = parameter.RefKind;
+				if( refKind == RefKind.None ) {
+
+					if( !serializedPropertyNames.Contains( parameter.Name ) ) {
+						ReportConstructorParameterCannotBeDeserialized( context, parameter );
+					}
+
+				} else {
+					ReportConstructorParameterInvalidRefKind( context, parameter, refKind );
+				}

```

Comment:
nit: I think we could raise both of these simultaneously, similar to raising diagnostics about the first constructor as well as the multiple constructors.

We've had some preference in our analyzers toward raising as many relevant diagnostics as possible (and not just the _first_ diagnostic), so that fixing a thing doesn't just proceed to then tell you additional things are wrong and you're just constantly re-building and playing whack-a-mole.

```suggestion
				if( !serializedPropertyNames.Contains( parameter.Name ) ) {
					ReportConstructorParameterCannotBeDeserialized( context, parameter );
				}

				if( parameter.RefKind == RefKind.None ) {
					ReportConstructorParameterInvalidRefKind( context, parameter, refKind );
				}
```